### ISSUE TITLE
Pattern Matching for switch (JEP406)

### ISSUE THREAD

**Issue by MysterAitch** [javaparser/javaparser#3328]
JDK17 - preview:
https://openjdk.java.net/jeps/406

> **Summary**
> 
> Enhance the Java programming language with pattern matching for switch expressions and statements, along with extensions to the language of patterns. Extending pattern matching to switch allows an expression to be tested against a number of patterns, each with a specific action, so that complex data-oriented queries can be expressed concisely and safely.
> 
> **Goals**
> 
> -  Expand the expressiveness and applicability of switch expressions and statements by allowing patterns to appear in case labels.
> -  Allow the historical null-hostility of switch to be relaxed when desired.
> -  Introduce two new kinds of patterns: guarded patterns, to allow pattern matching logic to be refined with arbitrary boolean expressions, and parenthesized patterns, to resolve some parsing ambiguities.
> -  Ensure that all existing switch expressions and statements continue to compile with no changes and execute with identical semantics.
> -  Do not introduce a new switch-like expression or statement with pattern-matching semantics that is separate from the traditional switch construct.
> -  Do not make the switch expression or statement behave differently when case labels are patterns versus when case labels are traditional constants.

Examples from the JEP: 

- standard

```java

static String formatterPatternSwitch(Object o) {
    return switch (o) {
        case Integer i -> String.format("int %d", i);
        case Long l    -> String.format("long %d", l);
        case Double d  -> String.format("double %f", d);
        case String s  -> String.format("String %s", s);
        default        -> o.toString();
    };
}
```

- checking for null

```java
static void testFooBar(String s) {
    switch (s) {
        case null         -> System.out.println("Oops");
        case "Foo", "Bar" -> System.out.println("Great");
        default           -> System.out.println("Ok");
    }
}
```

- "refining patterns in switch"

```java
static void testTriangle(Shape s) {
    switch (s) {
        case Triangle t && (t.calculateArea() > 100) ->
            System.out.println("Large triangle");
        case Triangle t ->
            System.out.println("Small triangle");
        default ->
            System.out.println("Non-triangle");
    }
}
```

**Comment by famod** [comment]
So, JDK 17 (latest LTS) is out for about a month now and javaparser seems to choke on something like:
```
final var principal = identity.getPrincipal();
if (principal instanceof final JsonWebToken jwt) {
```
```
Found "final", expected one of  "@" "boolean" "byte" "char" "double" "enum" "exports" "float" "int" "long" "module" "open" "opens" "provides" "record" "requires" "short" "strictfp" "to" "transitive" "uses" "with" "yield" <IDENTIFIER>
```
@MysterAitch shall I file a separate issue for this?

**Comment by mickroll** [comment]
@famod I think there is already a ticket, but sadly untouched for almost a year: https://github.com/javaparser/javaparser/issues/2445#issuecomment-964188096

**Comment by ice1000** [comment]
In JDK 18's second preview, it supports an enhanced coverage check based on generics.

**Comment by matozoid** [comment]
People, this is open source, you are all invited to help out!

**Comment by jlerbsc** [comment]
Pattern Matching for switch have been available since version 3.26.0.
